### PR TITLE
Changes needed to build with newer LLVM

### PR DESCRIPTION
**Pull request by arichardson**
Mostly cherry-picks from upstream. Only real change is the last commit that allows us to still use the old compiler which is > 13.0.0 but doesn't support -Wno-unused-but-set-variable yet.

### CODE REVIEW

**Comment by arichardson** [comment]
CI failure is the `vtmmio_attach()` purecap kernel tag violation again.

File: share/mk/bsd.compiler.mk [hunk 1]
```diff
@@ -244,7 +244,19 @@ ${X_}COMPILER_FEATURES=		c++11 c++14
 ${X_}COMPILER_FEATURES+=	c++17
 .endif
 .if ${${X_}COMPILER_TYPE} == "clang"
-${X_}COMPILER_FEATURES+=	retpoline init-all
```
Comment (brooksdavis) [inline share/mk/bsd.compiler.mk:247]: Is this removal intentional?

Comment (arichardson) [reply share/mk/bsd.compiler.mk:247]: No, will fix on Monday

**Review by brooksdavis — APPROVED** [review]

File: sys/conf/kern.mk [hunk 2]
```diff
@@ -36,7 +36,7 @@ CWARNEXTRA?=	-Wno-error=tautological-compare -Wno-error=empty-body \
 		-Wno-error=pointer-sign
 CWARNEXTRA+=	-Wno-error=shift-negative-value
 CWARNEXTRA+=	-Wno-address-of-packed-member
-.if ${COMPILER_VERSION} >= 130000
+.if ${COMPILER_VERSION} >= 130000 && ${COMPILER_FEATURES:MWunused-but-set-variable}
```
Comment (jrtc27) [inline sys/conf/kern.mk:39]: This is inconsistent, everywhere else replaces the version check. I'm in two minds which is best...

Comment (arichardson) [reply sys/conf/kern.mk:39]: Oh yeah, I initially added the check but then decided to replace and forgot to update this file. I don't mind either way but you are right I should be consistent.

**Comment by arichardson** [comment]
Will merge this later today/tomorrow morning unless there are any further objections.
